Added Cache Map (.basset)

promatik opened the pull request:
This PR aims the first possible improvement on https://github.com/Laravel-Backpack/basset/issues/18 👌

I'll start this by the results;

Previous loading times in _ms_ for [demo monsters edit page](https://demo.backpackforlaravel.com/admin/monster/140/edit), >200 bassets;
> 5.3005218505859  
> 4.9705505371094  
> 5.4311752319336  

Using cache map 🤯
> 1.5203952789307  
> 1.662015914917  
> 1.467227935791  

It's 3 times faster!!! 

---

There's a new config for enabling or disabling cacheMap `'cache_map' => true`.
In order to test the loading times, there's a _hidden_ config `'log_execution_time' => true` 👌

tabacitu commented on src/Helpers/Unarchiver.php:
```diff
@@ -0,0 +1,113 @@
+        return $phar->extractTo($output);
+    }
+
+    /**
+     * Returns a temporary file path.
+     *
+     * @return string
+     */
+    public function getTemporaryFilePath(): string
+    {
+        return tempnam(sys_get_temp_dir(), '');
+    }
+
+    /**
+     * Returns a temporary directory path.
+     *
+     * @return string
+     */
+    public function getTemporaryDirectoryPath(): string
+    {
+        $dir = storage_path('app/tmp/'.mt_rand().'/');
+        File::ensureDirectoryExists($dir);
+
+        return $dir;
+    }
```
Let's move this to the top of the file. It's easier to "read" a class if you start with `public` methods, then `private`. I also find it easier to read if you start with getters, then setters/doers.

tabacitu commented on src/Helpers/CacheMap.php:
```diff
@@ -0,0 +1,74 @@
+{
+    private $map = [];
+    private $path;
+    private $active = false;
+    private $dirty = false;
```
Maybe we add a verb here to make the call more fluent? I think I like `cachemap->isDirty` more than `cachemap->dirty()`, how about you?

```suggestion
    private $isActive = false;
    private $isDirty = false;
```

tabacitu commented on src/Helpers/CacheMap.php:
```diff
@@ -0,0 +1,74 @@
+     * @param  string  $path
+     * @return void
+     */
+    public function add(string $asset, string|bool $path = true): void
```
Hmm... I don't think `$cacheMap->add()` is very expressive... 

In previous PRs I was suggesting we remove the `asset` from the function name, because the `asset` was already present in the class name, but in this case, when you write `$cacheMap->add()`... I don't think it's very clear what you're adding...

This is nitpicking of course but I think it'd read more clearly if it were:

```suggestion
    public function addAsset(string $asset, string|bool $path = true): void
```

tabacitu commented on src/Helpers/CacheMap.php:
```diff
@@ -0,0 +1,74 @@
+     * @param  string  $asset
+     * @return string | false
+     */
+    public function get(string $asset): string|false
```
Similarly:

```suggestion
    public function getAsset(string $asset): string|false
```

tabacitu commented on src/Console/Commands/BassetClear.php:
```diff
@@ -42,6 +42,7 @@ public function handle(): void
 
         $disk->deleteDirectory($path);
         $disk->makeDirectory($path);
+        $disk->delete('.basset');
```
So this is something I wanted to talk to you about... Should we use the _singular_ name everywhere in the file paths... OR the _plural_? 👀 🤷‍♂️ Would it make more sense to be:
- `storage/app/public/basset` (or)
- `storage/app/public/bassets`

I mean... are we storing 
- "_things that Basset uses_" (or)
- "_backed up assets (aka bassets)_"

Which one sounds more intuitive to you?

promatik replied:
Good questions!
In the beggining we were using `bassets`, at some point (probably while creating the package) you named it `basset`, and from that point I started to call Basset to every of this "global" thigs.

I though that `.basset` is the file that stores Basset related stuff 🤷‍♂️
One day it may store more than just the map of bassets ...

`basset` is the folder named after the package name, that stores anything related with basset package ...

This is what I thought, `basset` works for everything, `bassets` only to mention assets "backed up assets" ...

Let me know if we should shift 😅

tabacitu replied:
Ok fine, I guess 😅 We can do a full code review after we know we're done. Maaaybe 😀

tabacitu commented on src/BassetManager.php:
```diff
@@ -30,6 +36,10 @@ public function __construct()
         $this->basePath = (string) Str::of(config('backpack.basset.path'))->finish('/');
 
+        $this->cacheMap = new CacheMap();
+        $this->loading = new LoadingTime();
```
Hmm I found this a bit weird... "loading" is not a noun so it's weird to do stuff to it, right? 😀

```suggestion
        $this->loader = new LoadingTime();
```

tabacitu commented on src/BassetManager.php:
```diff
@@ -153,29 +163,39 @@ public function getUrl(string $asset): string
      */
     public function basset(string $asset, bool|string $output = true, array $attributes = []): StatusEnum
     {
+        $this->loading->start();
```
Yup, there we go... this is a bit weird, right? Wouldn't it make more sense as:

```suggestion
        $this->loader->start();
```

tabacitu replied:
Or maybe:
```suggestion
        $this->loadingLog->start();
```

Something like that?

tabacitu reviewed (changes requested):
I love this! Let's do it! Left a few notes on class names etc - let me know what you think.

tabacitu commented:
Also, heads-up @promatik - please merge `main` into this branch to resolve the merge conflicts.

tabacitu commented on src/Helpers/CacheMap.php:
```diff
@@ -0,0 +1,77 @@
+     * @param  string  $path
+     * @return void
+     */
+    public function addAsset(string $asset, string|bool $path = true): void
```
I notice the assets get their ABSOLUTE paths added:

<img width="1512" alt="CleanShot 2023-04-21 at 11 45 54@2x" src="https://user-images.githubusercontent.com/1032474/233590288-d07e40ad-0b54-4d21-88a5-f0479c2f6a1b.png">

Doesn't this mean that... if something changes, like the protocol (HTTP/HTTPS) or app name (unlikely)... the whole bassets cache file is useless? Is that a real problem... or not? 🤷‍♂️

promatik replied:
Good question!
I noticed that, thought about and decided to keep it that way because the .basset file is local, so it's unlikely anything on the machine changes, and that is the default way, to remove the url would "cost" cpu time on each request to add the url again BUT it's irrelevant anyway.

I'll do it because it's safer that way 👌

promatik replied:
Done!

promatik commented:
1) I end up fixing more stuff in the `.basset` file, since the url is now removed, I also removed the path system path on basset block files and it's now ordered 🧹

![image](https://user-images.githubusercontent.com/1838187/234488876-231e2c24-83fd-463b-8f0b-5138cf272585.png)

2) Moved the `.basset` file inside the basset folder, it was in the root
`/storage/public/.basset` » `/storage/public/{$bassetPath = 'basset'}/.basset`.

3) All the variables have a types 🧹

4) GitHub CI is fixed 😅
